core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -137,12 +141,36 @@ class Partition(val topic: String,
     }
   }
 
+  /**
+    * Create the future replica if 1) the current replica is not in the given log directory and 2) the future replica
+    * does not exist. This method assumes that the current replica has already been created.
+    *
+    * @param logDir log directory
+    * @return true iff the future replica is created
+    */
+  def maybeCreateFutureReplica(logDir: String): Boolean = {
+    // The readLock is needed to make sure that while the caller checks the log directory of the
+    // current replica and the existence of the future replica, no other thread can update the log directory of the
+    // current replica or remove the future replica.
+    inReadLock(leaderIsrUpdateLock) {
+      val currentReplica = getReplica().get
+      if (currentReplica.log.get.dir.getParent == logDir)
+        false
+      else if (getReplica(Request.FutureLocalReplicaId).isDefined)
```

Comment:
Good point. I will update the patch to throw IllegalStateException.

Reply:
On a double thought, I realized that this could actually happen if the broker receives the same AlterReplicaLogDirsRequest multiple times, which could happen because ReassignPartitionsCommand will re-send AlterReplicaLogDirsRequest. Does this make sense?

Reply:
Hmm, this is the case that the future replica exists, but not in the specified log dir.  Resending the same AlterReplicaLogDirsRequest won't lead to this case, right?

Reply:
Not sure I explained my thought clearly. I think here the case is that the future replica already exists in the specified log dir. 

More specifically, when the broker receives the same AlterReplicaLogDirsRequest again, it calls ReplicaManager.alterReplicaLogDirs(). alterReplicaLogDirs() should only add this partition to replicaAlterLogDirsManager if this future replica has not already been created. So alterReplicaLogDirs() calls maybeCreateFutureReplica(), and maybeCreateFutureReplica() should return true iff the future replica is newly created. Because it is possible that future replica already exists, we should not throw IllegalStateException. Does this sound OK?

Reply:
@lindong28 : Thanks for the info. Makes sense. Perhaps we should throw an IllegalStateException if future replica exists, but is on a log dir different from the input?

Reply:
Sure. I have updated the code to throw IllegalStateException if the current log dir of the future replica is different from the requested log dir. Thanks!